src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -255,8 +256,57 @@ private void RecordDependency(CallGraphNode callerKey, CallGraphNode calledKey,
         /// </summary>
         public List<ImmutableArray<CallGraphNode>> GetCallCycles()
         {
-            // ToDo: need to implement a robust algorithm for finding cycles
-            throw new NotImplementedException();
+            var callStack = new Dictionary<CallGraphNode, CallGraphNode>();
+            var finished = new HashSet<CallGraphNode>();
+            var cycles = new List<ImmutableArray<CallGraphNode>>();
+
+            void processDependencies(CallGraphNode node)
+            {
+                if (_Dependencies.TryGetValue(node, out var dependencies))
+                {
+                    foreach (var temp in dependencies)
+                    {
+                        var (curr, _) = temp;
+
+                        if (!finished.Contains(curr))
+                        {
+                            if (callStack.ContainsKey(curr))
+                            {
+                                // Cycle detected
+
+                                var cycle = new List<CallGraphNode>() { curr };
+                                while (callStack.TryGetValue(curr, out var next))
+                                {
+                                    if (curr.Equals(next)) break;
+                                    cycle.Add(next);
+                                    curr = next;
+                                }
```

Comment:
Consider splitting this out into a separate (probably local) function like `List<CallGraphNode> getCallCycle(CallGraphNode node)`.

Reply:
It's not much code, and local functions can be confusing and clutter up the code.

Reply:
It doesn't have to be a local function. It can be a private method. :) IMO, having several small, well-named functions that are composed together enhances readability compared to one large function, especially since this section of code reassigns `curr`; the mutation of `curr` can be isolated to a function that is a few lines long, which makes it easier to reason about.